CA1001 should say which fields are disposable

#### Analyzer package

Example: Microsoft.CodeQuality.CSharp.Analyzers, version [2.6.0](https://www.nuget.org/packages/Microsoft.CodeQuality.CSharp.Analyzers/2.6.0)

#### Analyzer

CA1001: Types that own disposable fields should be disposable

#### Repro steps

Analyze this:
```C#
using System.IO;

namespace ClassLibrary1
{
    public class Class1
    {
        private readonly Stream _stream1 = new MemoryStream();
        private readonly Stream _stream2 = new MemoryStream();
        private readonly Stream _stream3 = new MemoryStream();
        private readonly FileInfo _fileInfo = new FileInfo(@"C:\test.txt");
    }
}

```

This reports CA1001 correctly, but I'm left to figure out on my own which fields are disposable. Even if I implement Dispose correctly, I have no confidence that I've called Dispose on all the disposable fields.

My suggestion is to augment CA1001's message with the name(s) of the disposable field(s).

#### Expected behavior
warning CA1001: Type 'Class1' owns disposable fields (_stream1, stream2, _stream3) but is not disposable

#### Actual behavior
warning CA1001: Type 'Class1' owns disposable fields but is not disposable

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx
@@ -177,11 +177,8 @@
   <data name="TypesThatOwnDisposableFieldsShouldBeDisposableDescription" xml:space="preserve">
     <value>A class declares and implements an instance field that is a System.IDisposable type, and the class does not implement IDisposable. A class that declares an IDisposable field indirectly owns an unmanaged resource and should implement the IDisposable interface.</value>
   </data>
-  <data name="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix" xml:space="preserve">
-    <value>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</value>
-  </data>
   <data name="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking" xml:space="preserve">
-    <value>Type '{0}' owns disposable fields but is not disposable</value>
+    <value>Type '{0}' owns disposable field(s) '{1}' but is not disposable</value>
   </data>
   <data name="UseGenericEventHandlerInstancesTitle" xml:space="preserve">
     <value>Use generic event handler instances</value>

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposable.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposable.cs
@@ -69,7 +69,7 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                     IEnumerable<IFieldSymbol> disposableFields = from member in namedType.GetMembers()
                                                                  where member.Kind == SymbolKind.Field && !member.IsStatic
                                                                  let field = member as IFieldSymbol
-                                                                 where field.Type != null && field.Type.AllInterfaces.Contains(_disposableTypeSymbol)
+                                                                 where field.Type != null && field.Type.IsDisposable(_disposableTypeSymbol)
                                                                  select field;
 
                     if (disposableFields.Any())
@@ -86,7 +86,10 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                                                                     symbolContext.CancellationToken));
                             if (syntaxNodes.Any())
                             {
-                                symbolContext.ReportDiagnostic(namedType.CreateDiagnostic(Rule, namedType.Name));
+                                // Type '{0}' owns disposable field(s) '{1}' but is not disposable
+                                var arg1 = namedType.Name;
+                                var arg2 = string.Join(", ", disposableFieldsHashSet.Select(f => f.Name).Order());
+                                symbolContext.ReportDiagnostic(namedType.CreateDiagnostic(Rule, arg1, arg2));
                                 return;
                             }
                         }

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.cs.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.cs.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Třída deklaruje a implementuje pole instance, které je typu System.IDisposable, přičemž tato třída neimplementuje IDisposable. Třída, která deklaruje pole IDisposable nepřímo, vlastní nespravovaný prostředek a měla by implementovat rozhraní IDisposable.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Implementujte IDisposable pro {0}, protože vytváří členy následujících typů IDisposable: {1}. Pokud byl typ {0} dříve součástí dodávky, považuje se přidání nových členů, které pro tento typ implementují IDisposable, za kritickou změnu pro existující konzumenty.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">Typ {0} vlastní jednoúčelová pole, ale není jednoúčelový.</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">Typ {0} vlastní jednoúčelová pole, ale není jednoúčelový.</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.de.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.de.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Eine Klasse deklariert und implementiert ein Instanzenfeld vom Typ "System.IDisposable", und IDisposable wird von der Klasse nicht implementiert. Eine Klasse, die ein IDisposable-Feld indirekt deklariert, besitzt eine nicht verwaltete Ressource und muss die IDisposable-Schnittstelle implementieren.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Implementieren Sie IDisposable für "{0}", weil dadurch Member der folgenden IDisposable-Typen erstellt werden: {1}. Wenn "{0}" zuvor geliefert wurde, wird das Hinzufügen neuer Member, die IDisposable auf diesen Typ anwenden, als fehlerhafte Änderung für vorhandene Endverbraucher betrachtet.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">Der Typ "{0}" besitzt verwerfbare Felder, ist jedoch nicht verwerfbar.</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">Der Typ "{0}" besitzt verwerfbare Felder, ist jedoch nicht verwerfbar.</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.es.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.es.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Una clase declara e implementa un campo de instancia que es un tipo System.IDisposable y la clase no implementa IDisposable. Una clase que declara un campo IDisposable indirectamente tiene un recurso no administrado y debe implementar la interfaz IDisposable.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Implemente IDisposable en {0} porque crea miembros de los tipos IDisposable siguientes: {1}. Si {0} ya se envió, agregar nuevos miembros que implementen IDisposable en este tipo se considera un cambio importante para los consumidores existentes.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">El tipo “{0}” tiene campos descartables, pero no es descartable</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">El tipo “{0}” tiene campos descartables, pero no es descartable</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.fr.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.fr.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Une classe déclare et implémente un champ d'instance qui représente le type System.IDisposable, et la classe n'implémente pas IDisposable. Une classe qui déclare un champ IDisposable est indirectement propriétaire d'une ressource non managée et doit implémenter l'interface IDisposable.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Implémenter IDisposable sur {0}, car il crée des membres des types IDisposable suivants : {1}. Si {0} a déjà été fourni, l'ajout de nouveaux membres qui implémentent IDisposable sur ce type est considéré comme une rupture pour les contrôles serveur consommateurs existants.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">Le type '{0}' a des champs supprimables mais n'est pas supprimable</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">Le type '{0}' a des champs supprimables mais n'est pas supprimable</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.it.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.it.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Una classe dichiara e implementa un campo di istanza che è un tipo System.IDisposable, ma non implementa IDisposable. Una classe che dichiara un campo IDisposable è indirettamente proprietaria di una risorsa non gestita e deve implementare l'interfaccia IDisposable.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Implementare IDisposable su {0} perché crea membri dei tipi IDisposable seguenti: {1}. Se {0} è stato distribuito in precedenza, l'aggiunta di nuovi membri che implementano IDisposable in questo tipo è considerata una modifica che determina un'interruzione per i consumer esistenti.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">Il tipo '{0}' è proprietario di campi Disposable ma non è Disposable</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">Il tipo '{0}' è proprietario di campi Disposable ma non è Disposable</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.ja.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.ja.xlf
@@ -102,14 +102,9 @@
         <target state="translated">クラスが System.IDisposable 型であるインスタンス フィールドを宣言および実装していますが、IDisposable を実装していません。IDisposable フィールドを宣言するクラスは間接的にアンマネージ リソースを所有しているため、IDisposable インターフェイスを実装する必要があります。</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">次の IDisposable 型のメンバーを作成するため、IDisposable を {0} で実装します: {1}。{0} が以前に出荷されている場合、IDisposable を実装する新しいメンバーを追加することは、既存のコンシューマーにとって重大な変更になります。</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">型 '{0}' は破棄可能なフィールドを所有しますが、破棄可能ではありません</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">型 '{0}' は破棄可能なフィールドを所有しますが、破棄可能ではありません</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.ko.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.ko.xlf
@@ -102,14 +102,9 @@
         <target state="translated">클래스는 System.IDisposable 형식인 인스턴스 필드를 선언하고 구현하지만, IDisposable을 구현하지 않습니다. IDisposable 필드를 선언하는 클래스는 관리되지 않는 리소스를 간접적으로 소유하며, IDisposable 인터페이스를 구현해야 합니다.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">IDisposable 형식 {1}의 멤버가 만들어지므로 {0}에 IDisposable을 구현하세요. {0}이(가) 이전에 제공된 경우 이 형식에 IDisposable을 구현하는 새 멤버를 추가하면 기존 소비자는 이를 새로운 변경 사항으로 인식합니다.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">삭제 가능한 필드가 있는 '{0}' 형식을 삭제할 수 없습니다.</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">삭제 가능한 필드가 있는 '{0}' 형식을 삭제할 수 없습니다.</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.pl.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.pl.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Klasa deklaruje i implementuje pole wystąpienia typu System.IDisposable oraz nie implementuje interfejsu IDisposable. Klasa deklarująca pole IDisposable jest pośrednio właścicielem zasobu niezarządzanego i powinna implementować interfejs IDisposable.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Zaimplementuj interfejs IDisposable dla klasy {0}, ponieważ tworzy ona elementy członkowskie następujących typów IDisposable: {1}. Jeśli klasa {0} została już dostarczona użytkownikom, dodanie nowych elementów członkowskich implementujących interfejs IDisposable do tego typu jest uznawane za zmianę, która może powodować błędy u obecnych użytkowników.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">Typ „{0}” zawiera pola możliwe do likwidacji, lecz sam nie jest możliwy do likwidacji</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">Typ „{0}” zawiera pola możliwe do likwidacji, lecz sam nie jest możliwy do likwidacji</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.pt-BR.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.pt-BR.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Uma classe declara e implementa um campo de instância que é um tipo System.IDisposable e a classe não implementa IDisposable. Uma classe que declara um campo IDisposable é proprietária indireta de um recurso não gerenciado e deve implementar a interface IDisposable.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Implemente IDisposable em {0} porque ele cria membros dos seguintes tipos IDisposable: {1}. Se {0} tiver sido enviado anteriormente, a adição de novos membros que implementem IDisposable a esse tipo será considerada uma alteração interruptiva para os consumidores existentes.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">O tipo '{0}' tem campos descartáveis, mas não é descartável</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">O tipo '{0}' tem campos descartáveis, mas não é descartável</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.ru.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.ru.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Класс объявляет и реализует поле экземпляра, имеющее тип System.IDisposable, и не реализует IDisposable. Класс, который объявляет поле IDisposable, косвенно владеет неуправляемым ресурсом и должен реализовывать интерфейс IDisposable.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">Реализуйте IDisposable для {0}, так как он создает члены следующих типов IDisposable: {1}. Если {0} ранее был доставлен, добавление в данный тип новых членов, реализующих IDisposable, считается критическим изменением для существующих потребителей.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">Тип "{0}" владеет высвобождаемыми полями, но не является высвобождаемым</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">Тип "{0}" владеет высвобождаемыми полями, но не является высвобождаемым</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.tr.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.tr.xlf
@@ -102,14 +102,9 @@
         <target state="translated">Bir sınıf, System.IDisposable türü olan bir örnek alanını bildirip uygular ve sınıf tarafından IDisposable uygulanmaz. IDisposable alanı bildiren bir sınıf dolaylı olarak yönetilmeyen bir kaynağa sahip olur ve IDisposable arabirimini uygulamalıdır.</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">{0} tarafından aşağıdaki IDisposable türlerinin üyeleri oluşturulduğundan bunun üzerinde IDisposable uygulayın: {1}. {0} daha önce gönderilmişse, bu türe IDisposable uygulayan yeni üyelerin eklenmesi, mevcut kullanıcılar için bozucu bir değişiklik olarak değerlendirilir.</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">'{0}' türü atılabilir alanlar içeriyor ancak atılabilir değil</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">'{0}' türü atılabilir alanlar içeriyor ancak atılabilir değil</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.zh-Hans.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.zh-Hans.xlf
@@ -102,14 +102,9 @@
         <target state="translated">一个类，它声明和实现 System.IDisposable 类型的实例字段，且它不实现 IDisposable。声明 IDisposable 字段的类间接拥有非托管资源，并且应该实现 IDisposable 接口。</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">在 {0} 上实现 IDisposable，因为它创建下列 IDisposable 类型的成员: {1}。如果以前提供过 {0}，对于现有使用者来说，向此类型添加用于实现 IDisposable 的新成员属于重大更改。</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">类型“{0}”具有可释放的字段，但该类型是不可释放的</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">类型“{0}”具有可释放的字段，但该类型是不可释放的</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.zh-Hant.xlf
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/xlf/MicrosoftApiDesignGuidelinesAnalyzersResources.zh-Hant.xlf
@@ -102,14 +102,9 @@
         <target state="translated">類別會宣告及實作類型為 System.IDisposable 的執行個體欄位，且該類別不會實作 IDisposable。宣告 IDisposable 欄位的類別會間接擁有非受控資源，且應實作 IDisposable 介面。</target>
         <note />
       </trans-unit>
-      <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageDependsOnFix">
-        <source>Implement IDisposable on {0} because it creates members of the following IDisposable types: {1}. If {0} has previously shipped, adding new members that implement IDisposable to this type is considered a breaking change to existing consumers.</source>
-        <target state="translated">因為 {0} 它會建立下列 IDisposable 類型的成員，所以請於其上實作 IDisposable: {1}。若先前已送出 {0}，則對這個類型新增實作 IDisposable 的新成員，會被視為現有消費者的中斷變更。</target>
-        <note />
-      </trans-unit>
       <trans-unit id="TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking">
-        <source>Type '{0}' owns disposable fields but is not disposable</source>
-        <target state="translated">類型 '{0}' 擁有可處置的欄位，但其本身並非可處置</target>
+        <source>Type '{0}' owns disposable field(s) '{1}' but is not disposable</source>
+        <target state="needs-review-translation">類型 '{0}' 擁有可處置的欄位，但其本身並非可處置</target>
         <note />
       </trans-unit>
       <trans-unit id="UseGenericEventHandlerInstancesTitle">

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposableTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposableTests.cs
@@ -57,7 +57,7 @@ using System.IO;
         FileStream newFile1, newFile2 = new FileStream(""data.txt"", FileMode.Append);
     }
 ",
-            GetCA1001CSharpResultAt(4, 18, "NoDisposeClass"));
+            GetCA1001CSharpResultAt(4, 18, "NoDisposeClass", "newFile1, newFile2"));
         }
 
         [Fact]
@@ -77,7 +77,7 @@ using System.IO;
         }
     }
 ",
-            GetCA1001CSharpResultAt(5, 18, "NoDisposeClass"));
+            GetCA1001CSharpResultAt(5, 18, "NoDisposeClass", "newFile"));
         }
 
         [Fact]
@@ -132,7 +132,7 @@ using System.IO;
         }
     }|]
 ",
-            GetCA1001CSharpResultAt(5, 18, "NoDisposeClass"));
+            GetCA1001CSharpResultAt(5, 18, "NoDisposeClass", "newFile"));
         }
 
         [Fact]
@@ -196,6 +196,24 @@ public class HasDisposeMethod : IDisposable
 ");
         }
 
+        [Fact, WorkItem(1562, "https://github.com/dotnet/roslyn-analyzers/issues/1562")]
+        public void CA1001CSharpTestWithIDisposableField()
+        {
+            VerifyCSharp(@"
+using System;
+using System.IO;
+
+namespace ClassLibrary1
+{
+    public class Class1
+    {
+        private readonly IDisposable _disp1 = new MemoryStream();
+    }
+}
+",
+            GetCA1001CSharpResultAt(7, 18, "Class1", "_disp1"));
+        }
+
         [Fact]
         public void CA1001BasicTestWithNoDisposableType()
         {
@@ -233,7 +251,7 @@ Imports System.IO
         Dim newFile As FileStream = New FileStream(""data.txt"", FileMode.Append)
     End Class
 ",
-            GetCA1001BasicResultAt(5, 18, "NoDisposeClass"));
+            GetCA1001BasicResultAt(5, 18, "NoDisposeClass", "newFile"));
 
             VerifyBasic(@"
 Imports System.IO
@@ -243,7 +261,7 @@ Imports System.IO
         Dim newFile1 As FileStream, newFile2 As FileStream = New FileStream(""data.txt"", FileMode.Append)
     End Class
 ",
-            GetCA1001BasicResultAt(5, 18, "NoDisposeClass"));
+            GetCA1001BasicResultAt(5, 18, "NoDisposeClass", "newFile1, newFile2"));
 
             VerifyBasic(@"
 Imports System.IO
@@ -254,7 +272,7 @@ Imports System.IO
         Dim newFile2 As FileStream = New FileStream(""data.txt"", FileMode.Append)
     End Class
 ",
-            GetCA1001BasicResultAt(5, 18, "NoDisposeClass"));
+            GetCA1001BasicResultAt(5, 18, "NoDisposeClass", "newFile1, newFile2"));
         }
 
         [Fact]
@@ -275,7 +293,7 @@ Imports System.IO
 
    End Class
 ",
-            GetCA1001BasicResultAt(6, 17, "NoDisposeMethod"));
+            GetCA1001BasicResultAt(6, 17, "NoDisposeMethod", "newFile"));
         }
 
         [Fact]
@@ -325,7 +343,7 @@ Imports System.IO
 
    End Class|]
 ",
-            GetCA1001BasicResultAt(5, 17, "NoDisposeMethod"));
+            GetCA1001BasicResultAt(5, 17, "NoDisposeMethod", "newFile"));
         }
 
         [Fact]
@@ -392,16 +410,32 @@ Imports System.IO
 ");
         }
 
-        private static DiagnosticResult GetCA1001CSharpResultAt(int line, int column, string objectName)
+        [Fact, WorkItem(1562, "https://github.com/dotnet/roslyn-analyzers/issues/1562")]
+        public void CA1001BasicTestWithIDisposableField()
         {
-            return GetCSharpResultAt(line, column, CSharpTypesThatOwnDisposableFieldsShouldBeDisposableAnalyzer.RuleId,
-                string.Format(MicrosoftApiDesignGuidelinesAnalyzersResources.TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking, objectName));
+            VerifyBasic(@"
+Imports System
+Imports System.IO
+
+Namespace ClassLibrary1
+    Class Class1
+        Private Readonly _disp1 As IDisposable = new MemoryStream()
+    End Class
+End Namespace
+",
+            GetCA1001BasicResultAt(6, 11, "Class1", "_disp1"));
         }
 
-        private static DiagnosticResult GetCA1001BasicResultAt(int line, int column, string objectName)
+        private static DiagnosticResult GetCA1001CSharpResultAt(int line, int column, string objectName, string disposableFields)
+        {
+            return GetCSharpResultAt(line, column, CSharpTypesThatOwnDisposableFieldsShouldBeDisposableAnalyzer.RuleId,
+                string.Format(MicrosoftApiDesignGuidelinesAnalyzersResources.TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking, objectName, disposableFields));
+        }
+
+        private static DiagnosticResult GetCA1001BasicResultAt(int line, int column, string objectName, string disposableFields)
         {
             return GetBasicResultAt(line, column, BasicTypesThatOwnDisposableFieldsShouldBeDisposableAnalyzer.RuleId,
-                string.Format(MicrosoftApiDesignGuidelinesAnalyzersResources.TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking, objectName));
+                string.Format(MicrosoftApiDesignGuidelinesAnalyzersResources.TypesThatOwnDisposableFieldsShouldBeDisposableMessageNonBreaking, objectName, disposableFields));
         }
     }
 }

--- a/src/Utilities/Extensions/ITypeSymbolExtensions.cs
+++ b/src/Utilities/Extensions/ITypeSymbolExtensions.cs
@@ -113,7 +113,7 @@ namespace Analyzer.Utilities.Extensions
         /// Indicates if the given <paramref name="type"/> is a reference type that implements <paramref name="iDisposable"/>.
         /// </summary>
         public static bool IsDisposable(this ITypeSymbol type, INamedTypeSymbol iDisposable)
-            => type.IsReferenceType && type.ImplementsIDisposable(iDisposable);
+            => type.IsReferenceType && (type == iDisposable || type.ImplementsIDisposable(iDisposable));
 
         public static IEnumerable<AttributeData> GetApplicableAttributes(this INamedTypeSymbol type)
         {

